File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int ans=0,csum=0;
        map<int,int> mp;
        int n=(int)nums.size();
        for(int i=0;i<n;++i){
            csum+=nums[i];
            if(mp.find(csum-k)!=mp.end()){
                ans+=mp[csum-k];
            }
            if(csum==k) ++ans;
            mp[csum]++;
        }
        return ans;
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer,Integer> freq= new HashMap<>();
        freq.put(0,1); int ans= 0, csum= 0;
        for(int i=0;i<nums.length;i++){
            csum+= nums[i];
            ans+= freq.getOrDefault(csum-k, 0);
            freq.put(csum, freq.getOrDefault(csum, 0)+1);
        }
        return ans;
    }
}